### PR TITLE
switch to g4 10.2p01

### PR DESCRIPTION
**Pull request by civanch**
Trying to switch Geant4 to 10.2p01 and CLHEP to 2.3.1.1 for the DEVEL branch only.

### CODE REVIEW

**Comment by davidlt** [comment]
This will break externals (ParFullCMS). I have this for my GCC 6.0.0 branch, there are some modifications that are needed to CLHEP first. I have this all compiling for GCC 6.0.0 without issues.

**Comment by civanch** [comment]
May be I should check if in the native G4 there were modifications and commit corresponding updates to ParFullSMS to this branch?

**Comment by ghost** [comment]
@davidlt Will you push those needed changes to CLHEP to its cms-externals repository?

**Comment by davidlt** [comment]
@civanch can you point me to the latest ParFullCMS code or/and repository? Would be very useful.
@degano I am will do this today.

**Comment by ghost** [comment]
@davidlt ParFullCMS has a repository in our externals: https://github.com/cms-externals/parfullcms.
Although it looks like we use the one from your repository in production: https://github.com/cms-sw/cmsdist/blob/IB/CMSSW_8_1_X/stable/geant4-parfullcms.spec.

**Comment by davidlt** [comment]
@degano it's not upstream, it's from a tarball I had (which was shared with me someone else). These copies of ParFullCMS are moving between few people. Instead I would like to find a true source of this package content.

**Comment by civanch** [comment]
@davidlt , there is a fresh copy of ParFullCMS which compiles to me standalone versus Geant4 10.2p01 and gcc493: ~vnivanch/public/geant4.10.2.p01/ParFullCMS

**Comment by davidlt** [comment]
It's still today (5 minutes till midnight). Found time and finally committed the required change to CLHEP. https://github.com/cms-externals/clhep/commit/87505534a473b0195037fae21575cbdbf53bfbbb

File: clhep.spec [hunk 1]
```diff
@@ -1,6 +1,6 @@
-### RPM external clhep 2.2.0.4
+### RPM external clhep 2.3.1.1
 
-%define tag 0e355c07fbee0b1c680073cc8d77e530a7edc6a9
+%define tag 91f5812f9ade2d61386bd34a2006160ccc0c2909
```
Comment (ghost) [inline clhep.spec]: @civanch Could you please change this tag to pick up the one corresponding to @davidlt submission? (87505534a473b0195037fae21575cbdbf53bfbbb)

**Comment by ghost** [comment]
please test

**Comment by davidlt** [comment]
Ah, true my SPEC looks a bit different. Please, change this change:

```
-rm %i/lib/libCLHEP-[A-Z]*-%realversion.$so
+rm -f %i/lib/libCLHEP-[A-Z]*-%realversion.$so
```

**Comment by civanch** [comment]
please test

**Comment by davidlt** [comment]
I have almost the same SPECs for GEANT4 and CLHEP for GCC 6.0.0 and no problems with Starlight.

**Comment by civanch** [comment]
From code inspection it seems that the code should compile - this variable is the member of the class.

**Comment by civanch** [comment]
Should we simply resubmit the test?

**Comment by davidlt** [comment]
I will test it locally in few hours (hopefully).

**Comment by davidlt** [comment]
Everything compiled just fine locally:

```
a/RPMS/slc6_amd64_gcc493
a/RPMS/slc6_amd64_gcc493/external+geant4-toolfile+1.0-cms2-1-1.slc6_amd64_gcc493.rpm
a/RPMS/slc6_amd64_gcc493/external+clhep+2.3.1.1-1-1.slc6_amd64_gcc493.rpm
a/RPMS/slc6_amd64_gcc493/external+geant4+10.02.p01-1-1.slc6_amd64_gcc493.rpm
```

**Comment by davidlt** [comment]
Try restarting the test.

**Comment by civanch** [comment]
please test

**Comment by civanch** [comment]
What I see is the build problem in generators, not in clhep or geant4.

**Comment by davidlange6** [comment]
caused by an interface change in CLHEP?

**Comment by civanch** [comment]
No, a see a pure c++ compilation warning disconnected with anything else. Likely situation was following: due to dependence on CLHEP this package was required to be recompiled and an error was uncovered with the current gcc.

**Comment by davidlt** [comment]
@davidlange6 unknown, it worked just fine for me twice on gcc530 and gcc600. One needs to look why it's failing in this particular setup or machine.

If I take `cms-sw:IB/CMSSW_8_1_X/next` merge `civanch:switch_to_g4.10.2p01` all works fine for me.

EDIT: ah, I have never checked Starlight compilation.

**Comment by davidlt** [comment]
Fixed and pushed changes to `IB/CMSSW_8_1_X/stable`. CLHEP newer versions require at least C++11, this is why it always worked on GCC 6.0.0 (C++14 is default).

The following was pushed:

```
diff --git a/starlight.spec b/starlight.spec
index be3f1ef..0796f72 100644
--- a/starlight.spec
+++ b/starlight.spec
@@ -21,7 +21,8 @@ export CLHEP_PARAM_PATH=${CLHEP_ROOT}
 cmake ../%{n}-%{realversion} \
  -DCMAKE_INSTALL_PREFIX:PATH="%{i}" \
  -DCMAKE_BUILD_TYPE=Realease \
- -DENABLE_CLHEP=ON
+ -DENABLE_CLHEP=ON \
+ -DCPP11=ON

 make %{makeprocesses} VERBOSE=1

```

Wait for an hour or so and restart the test.

**Comment by civanch** [comment]
please test

**Comment by davidlt** [comment]
If we believe the results, we have at least one issue:

```
*** glibc detected *** python: double free or corruption (fasttop): 0x0000000008186710 ***
======= Backtrace: =========
/lib64/libc.so.6[0x3e19075f4e]
/lib64/libc.so.6[0x3e19078cad]
/build/cmsbuild/jenkins-workarea/workspace/cmsdist-test-pr/testBuildDir/slc6_amd64_gcc493/external/gcc/4.9.3/lib64/libstdc++.so.6(_ZdlPv+0x18)[0x2acd83d8029f]
/build/cmsbuild/jenkins-workarea/workspace/cmsdist-test-pr/testBuildDir/slc6_amd64_gcc493/external/gcc/4.9.3/lib64/libstdc++.so.6(+0xdbd40)[0x2acd83df4d40]
/build/cmsbuild/jenkins-workarea/workspace/cmsdist-test-pr/testBuildDir/slc6_amd64_gcc493/external/gcc/4.9.3/lib64/libstdc++.so.6(_ZNSs4_Rep10_M_destroyERKSaIcE+0x51)[0x2acd83df437f]
/build/cmsbuild/jenkins-workarea/workspace/cmsdist-test-pr/CMSSW_8_1_DEVEL_X_2016-03-10-1100/external/slc6_amd64_gcc493/lib/libCLHEP-2.3.1.1.so(+0x10e3d2)[0x2acd9000c3d2]
/lib64/libc.so.6(exit+0xe2)[0x3e19035b22]
/lib64/libc.so.6(__libc_start_main+0x104)[0x3e1901ed64]
python[0x4006b1]
```

Again, have not seen such issues on GCC 6.0.0 personal builds.

Something is wrong here potentially with the testing itself.

```
*** glibc detected *** edmWriteConfigs: double free or corruption (fasttop): 0x0000000003877460 ***
======= Backtrace: =========
/lib64/libc.so.6[0x3e19075f4e]
/lib64/libc.so.6[0x3e19078cad]
/build/cmsbuild/jenkins-workarea/workspace/cmsdist-test-pr/testBuildDir/slc6_amd64_gcc493/external/gcc/4.9.3/lib64/libstdc++.so.6(_ZdlPv+0x18)[0x2b02311b329f]
/build/cmsbuild/jenkins-workarea/workspace/cmsdist-test-pr/testBuildDir/slc6_amd64_gcc493/external/gcc/4.9.3/lib64/libstdc++.so.6(+0xdbd40)[0x2b0231227d40]
/build/cmsbuild/jenkins-workarea/workspace/cmsdist-test-pr/testBuildDir/slc6_amd64_gcc493/external/gcc/4.9.3/lib64/libstdc++.so.6(_ZNSs4_Rep10_M_destroyERKSaIcE+0x51)[0x2b023122737f]
/cvmfs/cms-ib.cern.ch/week1/slc6_amd64_gcc493/cms/cmssw-patch/CMSSW_8_1_DEVEL_X_2016-03-10-1100/external/slc6_amd64_gcc493/lib/libCLHEP-2.2.0.4.so(+0x10a6e2)[0x2b0237be66e2]
/lib64/libc.so.6(exit+0xe2)[0x3e19035b22]
/lib64/libc.so.6(__libc_start_main+0x104)[0x3e1901ed64]
edmWriteConfigs[0x406d99]
```

It's clearly visible that it's using both version of CLHEP, which we shouldn't mix.

Just put it into DEVEL IB and see how it works, it just faster. DEVEL is our trash bin to test various things.

**Comment by smuzaffar** [comment]
agree on putting it in devel

**Comment by smuzaffar** [comment]
lets see how 23h00 IB goes.
cmsdist PRs should use full cmssw release instead of patch release to avoid mixing versions

**Comment by civanch** [comment]
Thanx for merging this PR to devil. Now it is clear that any test of Geant4 cannot work because a new dataset G4ENSDFSTATEDATA is required. It is a relatively small one... I will make a new issue.

**Pull request merged**